File: 380 Call Forwarding/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
	    StreamTokenizer st = new StreamTokenizer(System.in);
	    st.nextToken();
	    int N = (int)st.nval;
        StringBuilder output = new StringBuilder("CALL FORWARDING OUTPUT\n");
        for(int i = 1; i <= N; ++i) {
			output.append("SYSTEM " + i + "\n");
			ArrayList<Request> requests = new ArrayList<>();
			short source, time, duration, target;
			while(st.nextToken() == StreamTokenizer.TT_NUMBER && (source = (short)st.nval) != 0) {
			    st.nextToken();
			    time = (short)st.nval;
    		    st.nextToken();
    		    duration = (short)st.nval;
    		    st.nextToken();
    		    target = (short)st.nval;
    		    requests.add(new Request(source, time, duration, target));
    		}
    		while(st.nextToken() == StreamTokenizer.TT_NUMBER && (time = (short)st.nval) != 9000) {
    			ArrayList<Short> calls = new ArrayList<>();
    			st.nextToken();
    			calls.add((short)st.nval);
    			boolean deadEnd = false;
    			int j = 0;
    			while(j < requests.size() && !deadEnd) {
    				if(calls.get(calls.size() - 1) == requests.get(j).source) {
    					if(time >= requests.get(j).time && time <= requests.get(j).time + requests.get(j).duration) {
    						if(calls.contains(requests.get(j).target))
    							deadEnd = true;
    						else {
    							calls.add(requests.get(j).target);
    							j = 0;
    						}
    					} else
    						++j;
					} else
    					++j;
    			}
				output.append(String.format("AT %04d CALL TO %04d RINGS %04d\n", time, calls.get(0), deadEnd ? 9999 : calls.get(calls.size() - 1)));
    		}
    	}
	    System.out.print(output + "END OF OUTPUT\n");
	}
}

class Request {
	short source, time, duration, target;
	Request(short source, short time, short duration, short target) {
		this.source = source;
		this.time = time;
		this.duration = duration;
		this.target = target;
	}
}

File: 380 Call Forwarding/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
	short source, time, duration, target;
} Request;

typedef struct {
	short extention;
} Call;

int main() {
	int N;
	scanf("%d", &N);
	puts("CALL FORWARDING OUTPUT");
	for(int i = 1; i <= N; ++i) {
		printf("SYSTEM %d\n", i);
		Request *requests = NULL;
		int requestCount = 0;
		short source;
		while(scanf("%hd", &source) && source != 0) {
			requests = (Request*)realloc(requests, (requestCount + 1) * sizeof(Request));
			requests[requestCount].source = source;
			scanf("%hd %hd %hd", &requests[requestCount].time, &requests[requestCount].duration, &requests[requestCount].target);
			++requestCount;
		}
		short time;
		while(scanf("%hd", &time) && time != 9000) {
		    Call *calls = NULL;
		    calls = (Call*)realloc(calls, sizeof(Call));
			scanf("%hd", &calls[0].extention);
			bool deadEnd = false;
			int j = 0, callCount = 1;
			while(j < requestCount && !deadEnd) {
				if(requests[j].source == calls[callCount - 1].extention) {
					if(time >= requests[j].time && time <= requests[j].time + requests[j].duration) {
						int k = 0;
						while(k < callCount && calls[k].extention != requests[j].target)
							++k;
						if(k < callCount)
							deadEnd = true;
						else {
							calls = (Call*)realloc(calls, (callCount + 1) * sizeof(Call));
							calls[callCount++].extention = requests[j].target;
							j = 0;
						}
					} else
						++j;
				} else
                    ++j;
			}
			printf("AT %04hd CALL TO %04hd RINGS %04hd\n", time, calls[0].extention, deadEnd ? 9999 : calls[callCount - 1].extention);
			free(calls);
		}
		free(requests);
	}
	puts("END OF OUTPUT");
	return 0;
}
